src/utils/url.js

```diff
@@ -31,6 +31,20 @@ export const encodeParamsForUrl = (params: UrlParams): string =>
     .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value.toString())}`)
     .join('&');
 
+/**
+ * Turn a relative or absolute URL into an absolute URL.
+ *
+ * If it starts with "http", it's assumed to be an absolute URL
+ * already and is returned without changes.
+ *
+ * Otherwise, it's a relative URL and is appended to the realm,
+ * deduplicating a leading slash if necessary.
+ *
```

Comment:
Some doc-writing style: these two sentences are really describing the current implementation, not the intended interface. The interface is what it's meant to do, from the perspective of writing code that uses it. Docs should describe the interface.

One way to see the difference is to notice that the behavior these sentences describe (which indeed is what the implementation does) is buggy, mishandling cases like `http?foo#bar` which are perfectly good relative URLs. The fact we can say it's "buggy" means the intended behavior is something else; it's buggy because it doesn't match that intended behavior.

(More often, the two don't disagree like they do here, but just describe different aspects: the interface is about expectations, and how the code relates to other code and to things outside the code.)

The fact that this implementation doesn't really match is good to write down too! Could be a line in the jsdoc, or just a `TODO` comment (outside the jsdoc, as `//` after it.) Probably enough to say "uses some crude heuristics" rather than go into detail, which rapidly winds up just repeating what the code says.

Reply:
Mm, thanks for catching and explaining this! This was my bad. When I recommended this JSDoc at https://github.com/zulip/zulip-mobile/pull/4089#discussion_r424652743, I was reminded of what you said at https://github.com/zulip/zulip-mobile/pull/4046#issuecomment-618044109; that change was merged as 452edbc5e. I think this part stuck in my mind:

> facing both inward to the implementation and outward to the call sites, of what they should rely on and what would be a bug

because it said why it was important to write something down instead of nothing, in cases like these where things are buggy / messy. The "Docs should describe the interface" (as opposed to the implementation) direction was clear, in that case, and I seem to have followed it there — but I forgot that part when I recommended this JSDoc on this PR; sorry about that. 🙂

Reply:
Okay .I've removed the implementation details and moved the TODO to a comment below the JSDoc.